File: Loom.Engine/Infrastructure/ExecutionRouter.vb
Imports Loom.Core.Interfaces
Imports Loom.Core.Models

Namespace Infrastructure
    ''' <summary>
    ''' Records new adapter and routes user's input model to a specific provider
    ''' </summary>
    Public Class ExecutionRouter
        Private ReadOnly _adapters As New List(Of IProviderAdapter)
        Private ReadOnly _catalog As New ModelCatalog()

        Public Sub RegisterAdapter(adapter As IProviderAdapter)
            If Not _adapters.Any(Function(a) a.ProviderName = adapter.ProviderName) Then
                _adapters.Add(adapter)
            End If
        End Sub

        ''' <summary>
        ''' Selects the provider based on the entered model
        ''' </summary>
        ''' <param name="invocation"></param>
        ''' <returns>Selected provider</returns>
        Public Function Route(invocation As ILlmInvocation) As IProviderAdapter

            Dim targetModelId = invocation.Hints.PreferredModel
            Dim selectedModelInfo As ModelMetaData = Nothing

            If Not String.IsNullOrWhiteSpace(targetModelId) Then
                selectedModelInfo = _catalog.GetModelInfo(targetModelId)
            End If

            If selectedModelInfo Is Nothing Then
                selectedModelInfo = _catalog.GetFallBack(invocation.Hints.Priority)

                invocation.Hints.PreferredModel = selectedModelInfo.ModelId
            End If

            Dim adapter = _adapters.FirstOrDefault(Function(a) a.ProviderName = selectedModelInfo.ProviderName)

            If adapter Is Nothing Then
                Throw New Exception($"No adapter installed for the provider: {selectedModelInfo.ProviderName}")
            End If

            Return adapter
        End Function

    End Class

End Namespace

File: implementations/csharp/src/Loom.Engine/Infrastructure/ExecutionRouter.cs
using Loom.Core.Interfaces;
using Loom.Core.Models;
using System.Collections.Generic;
using System.Linq;

namespace Loom.Engine.Infrastructure
{
    /// <summary>
    /// Records new adapter and routes user's input model to a specific provider
    /// </summary>
    public class ExecutionRouter
    {

        private readonly List<IProviderAdapter> _adapters = new List<IProviderAdapter>();
        private readonly ModelCatalog _catalog = new ModelCatalog();


        public void RegisterAdapter(IProviderAdapter adapter)
        {
            if (!_adapters.Any(a => a.ProviderName == adapter.ProviderName))
                _adapters.Add(adapter);
        }

        /// <summary>
        /// Selects the provider based on the entered model
        /// </summary>
        /// <param name="invocation"></param>
        /// <returns>Selected provider</returns>
        public IProviderAdapter Route(ILlmInvocation invocation)
        {
            var targetModelId = invocation.Hints.PreferredModel;
            ModelMetaData selectedModelInfo = null;

            if (!string.IsNullOrWhiteSpace(targetModelId))
                selectedModelInfo = _catalog.GetModelInfo(targetModelId);

            if (selectedModelInfo == null)
            {
                selectedModelInfo = _catalog.GetFallBack(invocation.Hints.Priority);
                invocation.Hints.PreferredModel = selectedModelInfo.ModelId;
            }

            var adapter = _adapters.FirstOrDefault(a => a.ProviderName == selectedModelInfo.ProviderName);

            if (adapter == null)
                throw new Exception($"No adapter installed for the provider: {selectedModelInfo.ProviderName}");

            return adapter;
        }

    }
}